File: apps/mobile/ios/App/App/MicPermissionPlugin.swift
// Fills a gap in Capacitor's OWN iOS WKUIDelegate: `WebViewDelegationHandler.swift`'s
// `requestMediaCapturePermissionFor` always calls `decisionHandler(.grant)` unconditionally at the app
// layer, so there is no control point before WebKit hits the real OS-level microphone permission — and
// iOS never re-shows that OS dialog once a user has denied it once. `useVoiceRecording.ts`'s
// `getUserMedia` just keeps rejecting with `NotAllowedError`, forever, with no way for the user to know
// why or how to recover, since neither `AppDelegate.swift` nor `MainViewController.swift` has any
// microphone-permission code at all. Unlike Android's equivalent plugin, `AVCaptureDevice`'s own
// `authorizationStatus` is always authoritative — no bookkeeping needed to distinguish "never asked"
// from "denied", the actual bug that broke Android's first attempt at this (see its own doc comment).
// `request()` still exists as its own method (not just `check()`) so both platforms share the exact
// same JS-side call shape (`useVoiceRecording.ts` calls `request()`, not `check()`, right before trying
// to record) — and so the OS dialog fires at a moment the app controls explicitly, same as Android,
// rather than implicitly whenever WebKit's own internal `getUserMedia` handling happens to trigger it.
import Capacitor
import AVFoundation
import UIKit

@objc(MicPermissionPlugin)
public class MicPermissionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MicPermissionPlugin"
    public let jsName = "MicPermission"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "check", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "request", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openSettings", returnType: CAPPluginReturnPromise),
    ]

    @objc func check(_ call: CAPPluginCall) {
        call.resolve(stateResult())
    }

    @objc func request(_ call: CAPPluginCall) {
        if AVCaptureDevice.authorizationStatus(for: .audio) == .notDetermined {
            AVCaptureDevice.requestAccess(for: .audio) { _ in
                DispatchQueue.main.async {
                    call.resolve(self.stateResult())
                }
            }
        } else {
            call.resolve(stateResult())
        }
    }

    private func stateResult() -> [String: Any] {
        let state: String
        switch AVCaptureDevice.authorizationStatus(for: .audio) {
        case .authorized:
            state = "granted"
        case .denied, .restricted:
            state = "blocked"
        case .notDetermined:
            fallthrough
        @unknown default:
            state = "prompt"
        }
        return ["state": state]
    }

    @objc func openSettings(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let url = URL(string: UIApplication.openSettingsURLString) else {
                call.reject("Couldn't build the Settings URL")
                return
            }
            UIApplication.shared.open(url, options: [:]) { _ in call.resolve() }
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/veasnawt/vcut/MicPermissionPlugin.kt
package com.veasnawt.vcut

import android.Manifest
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import androidx.core.app.ActivityCompat
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback

private const val MIC_ALIAS = "microphone"

/**
 * Fills a gap in Capacitor's OWN Android permission handling: `BridgeWebChromeClient.onPermissionRequest`
 * (vendored inside `@capacitor/android`, not this app's code) unconditionally re-launches the RECORD_AUDIO
 * runtime-permission request on every `getUserMedia` call via its OWN separate `ActivityResultContracts`
 * launcher, with no `shouldShowRequestPermissionRationale` check first — and critically, that launcher is
 * NOT the same one `Plugin.requestPermissionForAlias` uses below, so Capacitor's own permission-state
 * cache (`Bridge`'s `PluginPermStates` prefs) never learns about denials that happen through it either.
 * Once Android permanently denies the permission (its own policy after enough denials), that WebView call
 * just silently returns denied with no dialog at all — `useVoiceRecording.ts`'s `getUserMedia` rejects with
 * `NotAllowedError` again, forever, with no way for the user to know WHY or what to do next.
 *
 * First attempt at this plugin tracked "have we ever asked" in its own SharedPreferences flag, reasoning
 * `shouldShowRequestPermissionRationale` alone can't distinguish "never asked" from "permanently denied"
 * (both return `false`). That broke in real testing: a debug build installed OVER an existing copy of the
 * app (not a clean reinstall) inherits the OS's real denial history, but starts with a FRESH SharedPreferences
 * flag — so `check()` always read "never asked" and never reported "blocked", even once the OS had already
 * permanently denied it. Fixed by not tracking anything ourselves at all: `request()` below performs a REAL
 * permission request through Capacitor's own `requestPermissionForAlias`/`getPermissionState` machinery —
 * calling the actual Android API this way means `shouldShowRequestPermissionRationale`, read immediately
 * after, reflects the OS's true history (which IS tracked system-side across every past caller, including
 * the WebView's own separate launcher above), not anything this plugin remembers on its own. `check()`
 * still can't fully disambiguate "never asked at all, ever" from "blocked" before the FIRST real `request()`
 * this plugin ever makes — an unavoidable Android limitation — so callers should treat `check()` as a
 * best-effort hint for showing UI before the first tap, and always call `request()` (not just rely on
 * `check()`) right before actually trying to record, since only `request()`'s result is authoritative.
 */
@CapacitorPlugin(name = "MicPermission", permissions = [Permission(alias = MIC_ALIAS, strings = [Manifest.permission.RECORD_AUDIO])])
class MicPermissionPlugin : Plugin() {
    @PluginMethod
    fun check(call: PluginCall) {
        call.resolve(stateResult())
    }

    @PluginMethod
    fun request(call: PluginCall) {
        if (getPermissionState(MIC_ALIAS) == PermissionState.GRANTED) {
            call.resolve(stateResult())
            return
        }
        requestPermissionForAlias(MIC_ALIAS, call, "onMicPermissionResult")
    }

    @PermissionCallback
    private fun onMicPermissionResult(call: PluginCall) {
        call.resolve(stateResult())
    }

    private fun stateResult(): JSObject {
        val state = when {
            getPermissionState(MIC_ALIAS) == PermissionState.GRANTED -> "granted"
            ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.RECORD_AUDIO) -> "prompt"
            else -> "blocked"
        }
        val result = JSObject()
        result.put("state", state)
        return result
    }

    @PluginMethod
    fun openSettings(call: PluginCall) {
        try {
            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
            intent.data = Uri.fromParts("package", context.packageName, null)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
            call.resolve()
        } catch (e: Exception) {
            call.reject("Couldn't open Settings", e)
        }
    }
}
